File: cloudflare/src/brand-match.ts
import type { ObservationState } from "./types.ts";

const MAX_PATTERN_LENGTH = 200;
const MAX_RESPONSE_LENGTH = 200_000;
const MAX_MATCHES = 10;

export interface BrandMatchResult {
  state: ObservationState;
  matches: string[];
}

export function compileBrandRegex(pattern?: string) {
  const value = String(pattern || "").trim();
  if (!value) return undefined;
  if (value.length > MAX_PATTERN_LENGTH) {
    throw new Error(`Brand regex must be ${MAX_PATTERN_LENGTH} characters or fewer.`);
  }
  try {
    return new RegExp(value, "giu");
  } catch (error) {
    throw new Error(
      `Brand regex is invalid: ${error instanceof Error ? error.message : "invalid expression"}`,
    );
  }
}

export function matchBrand(
  responseText: string | undefined,
  pattern: string | undefined,
): BrandMatchResult {
  const regex = compileBrandRegex(pattern);
  if (!regex) return { state: "N/A", matches: [] };
  if (!responseText) return { state: "UNKNOWN", matches: [] };

  const matches: string[] = [];
  const seen = new Set<string>();
  for (const match of responseText.slice(0, MAX_RESPONSE_LENGTH).matchAll(regex)) {
    const value = match[0];
    const key = value.toLocaleLowerCase();
    if (value && !seen.has(key)) {
      seen.add(key);
      matches.push(value);
    }
    if (matches.length >= MAX_MATCHES) break;
  }
  return { state: matches.length ? "YES" : "NO", matches };
}

File: core/brand_match.py
from __future__ import annotations

import re

from .enums import ObservationState

MAX_PATTERN_LENGTH = 200
MAX_RESPONSE_LENGTH = 200_000
MAX_MATCHES = 10


def compile_brand_regex(pattern: str | None) -> re.Pattern[str] | None:
    value = str(pattern or "").strip()
    if not value:
        return None
    if len(value) > MAX_PATTERN_LENGTH:
        raise ValueError(f"Brand regex must be {MAX_PATTERN_LENGTH} characters or fewer.")
    try:
        return re.compile(value, re.IGNORECASE | re.UNICODE)
    except re.error as exc:
        raise ValueError(f"Brand regex is invalid: {exc}") from exc


def match_brand(
    response_text: str | None,
    pattern: str | None,
) -> tuple[ObservationState, list[str]]:
    regex = compile_brand_regex(pattern)
    if not regex:
        return ObservationState.NOT_APPLICABLE, []
    if not response_text:
        return ObservationState.UNKNOWN, []

    matches: list[str] = []
    seen: set[str] = set()
    for match in regex.finditer(response_text[:MAX_RESPONSE_LENGTH]):
        value = match.group(0)
        key = value.casefold()
        if value and key not in seen:
            seen.add(key)
            matches.append(value)
        if len(matches) >= MAX_MATCHES:
            break
    return (ObservationState.YES if matches else ObservationState.NO), matches
